bfvmm/include/hve/phys_pci.h

```diff
@@ -0,0 +1,370 @@
+    /**
+     * @brief Set the contents of the specified base address register.
+     *
+     * @expects n < 6
+     * @ensures invalid indices cause no write
+     *
+     * @param n index of the desired base address register
+     * @param value value to write
+     */
+    inline void set_bar(uint8_t n, uint32_t value)
+    {
+        if (n < 6) {
+            write_register(0x10 + 4 * n, value);
+        }
+    }
+
+    /**
+     * @brief Read 8 bits from a register.
+     *
+     * @expects reg <= 0x3C
+     * @ensures
+     *
+     * @param reg register index in bytes
+     */
+    inline uint8_t read_register_u8(register_type reg) const
+    {
+        return read_register_u8(m_bus, m_device, m_func, reg);
+    }
+
+    /**
+     * @brief Read 16 bits from a register.
+     *
+     * @expects reg <= 0x3C and reg is 16-bit aligned
+     * @ensures
+     *
+     * @param reg register index in bytes
+     */
+    inline uint16_t read_register_u16(register_type reg) const
+    {
+        return read_register_u16(m_bus, m_device, m_func, reg);
+    }
+
+    /**
+     * @brief Read 32 bits from a register.
+     *
+     * @expects reg <= 0x3C and reg is 32-bit aligned
+     * @ensures
+     *
+     * @param reg register index in bytes
+     */
+    inline uint32_t read_register_u32(register_type reg) const
+    {
+        return read_register_u32(m_bus, m_device, m_func, reg);
+    }
+
+    /**
+     * @brief Write 32 bits to a register.
+     *
+     * @expects reg <= 0x3C and reg is 32-bit aligned
+     * @ensures
+     *
+     * @param reg register index in bytes
+     * @param val value to write
+     */
+    void write_register(register_type reg, uint32_t val)
+    {
+        write_register(m_bus, m_device, m_func, reg, val);
+    }
+
+    /**
+     * @brief Perform an 8-bit read-modify-write operation on a 32-bit register.
+     *
+     * @expects reg <= 0x3C
+     * @ensures
+     *
+     * @param reg register index in bytes
+     * @param val value to write
+     */
+    void rmw_register_u8(register_type reg, uint8_t val)
+    {
+        rmw_register_u8(m_bus, m_device, m_func, reg, val);
+    }
+
+    /**
+     * @brief Perform a 16-bit read-modify-write operation on a 32-bit register.
+     *
+     * @expects reg <= 0x3C and reg is 16-bit aligned
+     * @ensures
+     *
+     * @param reg register index in bytes
+     * @param val value to write
+     */
+    void rmw_register_u16(register_type reg, uint16_t val)
+    {
+        rmw_register_u16(m_bus, m_device, m_func, reg, val);
+    }
+
+    /**
+     * @brief Read 8 bits from a geographically addressed register.
+     *
+     * @expects bus <= 255, device <= 31, func <= 7, reg <= 0x3C
+     * @ensures
+     *
+     * @param bus PCI bus number
+     * @param device Device number on the bus
+     * @param func Function, for multifunction devices (or 0)
+     * @param reg register index in bytes
+     */
+    static uint8_t read_register_u8(bus_type bus, device_type device,
```

Comment:
Please move the static functions to an intrinsics file, similar to lapic_register.{h,cpp}. You could put them in `namespace pci` or something similar

Reply:
Isn't that what I proposed in the original RFC, that you said not to do? :P

will do.